src/crimson/osd/pg.cc

```diff
@@ -146,6 +146,28 @@ PG::PG(
 
 PG::~PG() {}
 
+void PG::check_blocklisted_watchers()
+{
+  logger().debug("{}", __func__);
+  shard_services.for_each_obc([this](ObjectContextRef obc) {
```

Comment:
`for_each_obc` traverses through `obc_lru` which includes only cached obc (up to `crimson_osd_obc_lru_size`).
If we need to `for_each` all the existing obc, the ones which are stored in disc should also be taken into account.
This may cause excessive disc IO's just for checking blocklisted watchers.

Reply:
Classic corresponding implementation uses `while (object_contexts.get_next`).
The key difference is that classic's `SharedLRU` can return objects even after they have been evicted from cache using `weak_refs` map.

See header comment:
```
/// SharedLRU does its best to cache objects. It not only tracks the objects
/// in its LRU cache with strong references, it also tracks objects with
/// weak_ptr even if the cache does not hold any strong references to them. so
/// that it can return the objects after they are evicted, as long as they've
/// ever been cached and have not been destroyed yet.
```
In Crimson we use `intrusive_lru` which AFAIU doesn't support weak refs (yet).
```
  void evict() {
    while (!unreferenced_list.empty() &&
	   lru_set.size() > lru_target_size) {
      auto &b = unreferenced_list.front();
      assert(!b.lru);
      unreferenced_list.pop_front();
      lru_set.erase_and_dispose(
	lru_set.iterator_to(b),
	[](auto *p) { delete p; }
      );
    }
  }
```

Reply:
Working on it. I'm afraid that missing the _registry_ behavior might be far more problematic that just the client blocking per se. I wonder about a situation in which multiple `ObjectContext` for the same object are used the same time.

Reply:
We're evicting only from the `unreferenced_list`. An instance is being put on it when the last smart pointer is taken out:

```cpp
template <typename Config>
void intrusive_ptr_release(intrusive_lru_base<Config> *p) {
  assert(p);
  assert(p->use_count > 0);
  --p->use_count;
  if (p->use_count == 0) {
    p->lru->unreferenced(*p);
  }
}
```

So, we should be good -- it's a registry, not only a cache.

Reply:
> So, we should be good -- it's a registry, not only a cache.

Yes, assuming we keep referencing relevant obcs.

I pushed #49779 for clarification.